Update TypeScript SDK reference (#181)
clockworklabs/spacetime-docs: docs/sdks/typescript/index.md

mkdir -p client/src/module_bindings
spacetime generate --lang typescript \
    --out-dir client/src/module_bindings \
    --project-path server
```

And now you will get the files for the `reducers` & `tables`:

```bash
quickstart-chat
├── client
│   ├── node_modules
│   ├── public
│   └── src
|       └── module_bindings
|           ├── add_reducer.ts
|           ├── person.ts
|           └── say_hello_reducer.ts
└── server
    └── src
    --project-path PATH-TO-MODULE-DIRECTORY
```

Import the `module_bindings` in your client's _main_ file:

```typescript
import { SpacetimeDBClient, Identity } from '@clockworklabs/spacetimedb-sdk';

> import Person from './module_bindings/person';

Reviewer: NIT: Sometimes you do this kind of import, but it's not that typical, and sometimes discouraged. More typical to do: 

```ts
import { DbConnection } from './module_bindings/index'
```

But if we do an import the variable we bind it to should follow TypeScript casing rules, so: `moduleBindings` or just `module`.

Author: > `import { DbConnection } from './module_bindings/index'`

We don't want to do the named import style of `import { stuff... } from path` because:
- We have to import a whole lot of stuff from the module bindings.
- In this document, we don't have a specific module that we're writing against, so we *can't* list all of the items that we'd need to import. E.g. what would the list of row types be?

> But if we do an import the variable we bind it to should follow TypeScript casing rules, so: `moduleBindings` or just `module`.

It seems weird to me to have the file name not match the imported variable name, and we've pretty firmly committed to using `module_bindings` as the directory name in all our TypeScript examples. But sure, if you think this is better.

> | Name                                                      | Description                                                                                      |

Reviewer: This is a table of contents, right? Maybe we should call this "Section" instead of "Name"?

Author: Honestly, we could just remove the table header. But I'd prefer to do that later; for now, we're using the `| Name | Description |` formatting in several places, and it's what's listed in the style guide.

Reviewer: Okay. I'll just resolve this comment.

```typescript

> DbConnection.builder(): DbConnectionBuilder

Reviewer: ```suggestion
builder(): DbConnectionBuilder
```

Or

```ts
class DbConnectionBuilder {
  public withUri(uri: string): DbConnectionBuilder;
}

If you prefer. The above implies that it is a class method.

Reviewer: Let's do the latter I guess.

```suggestion
class DbConnection {
  public builder(): DbConnectionBuilder;
}
```

Author: I see what you mean. This one *is* what I would call a static method, or an associated function, which takes the class as a receiver. But I used this same syntax for what I would call methods, which take an instance as a receiver. What syntax would you use for each of these things, in general?

Reviewer: Oh, whoops, you're right. It's fine for this particular one, but confusing for the others. For static what you have here is fine:

```ts
DbConnection.builder(): DbConnectionBuilder
```

For methods let's do this:

```ts
class DbConnectionBuilder {
  public withUri(uri: string): DbConnectionBuilder
}
```

Author: If we're going to use the `class` syntax, let's write static methods as static methods, [since JS apparently supports that](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static).

```

> Construct a `DbConnection` by calling `DbConnection.builder()` and chaining configuration methods, then calling `.build()`. You must at least specify `withUri`, to supply the URI of the SpacetimeDB to which you published your module, and `with_module_name`, to supply the human-readable SpacetimeDB domain name or the raw `Identity` which identifies the module.

Reviewer: ```suggestion
Construct a `DbConnection` by calling `DbConnection.builder()` and chaining configuration methods, then calling `.build()`. You must at least specify `withUri`, to supply the URI of the SpacetimeDB to which you published your module, and `withModuleName`, to supply the human-readable SpacetimeDB domain name or the raw `Identity` which identifies the module.
```

|-------------------------------------------------------|--------------------------------------------------------------------------------------|
| [`withUri` method](#method-withuri)                   | Set the URI of the SpacetimeDB instance which hosts the remote database.             |
| [`withModuleName` method](#method-withmodulename)     | Set the name or `Identity` of the remote database.                                   |
| [`onConnect` callback](#callback-onconnect)           | Register a callback to run when the connection is successfully established.          |
| [`onConnectError` callback](#callback-onconnecterror) | Register a callback to run if the connection is rejected or the host is unreachable. |
| [`onDisconnect` callback](#callback-ondisconnect)     | Register a callback to run when the connection ends.                                 |
| [`withToken` method](#method-withtoken)               | Supply a token to authenticate with the remote database.                             |
| [`build` method](#method-build)                       | Finalize configuration and connect.                                                  |

#### Method `withUri`

```typescript

> DbConnectionBuilder.withUri(uri: string): DbConnectionBuilder

Reviewer: ```suggestion
class DbConnectionBuilder {
  public withUri(uri: string): DbConnectionBuilder
}
```

```

Configure the URI of the SpacetimeDB instance or cluster which hosts the remote module.

#### Method `withModuleName`

```typescript

> DbConnectionBuilder.withModuleName(name_or_identity: string): DbConnectionBuilder

Reviewer: Probably easier to update this in a proper editor rather than adding suggestions, but ditto on all of them.

```

> Configure the SpacetimeDB domain name or `Identity` of the remote module which identifies it within the SpacetimeDB instance or cluster.

Reviewer: ```suggestion
Configure the SpacetimeDB domain name or hex string encoded `Identity` of the remote module which identifies it within the SpacetimeDB instance or cluster.
```

#### Callback `onConnect`

```typescript
DbConnectionBuilder.onConnect(
    callback: (ctx: DbConnection, identity: Identity, token: string) => void
): DbConnectionBuilder
```

Chain a call to `.onConnect(callback)` to your builder to register a callback to run when your new `DbConnection` successfully initiates its connection to the remote module. The callback accepts three arguments: a reference to the `DbConnection`, the `Identity` by which SpacetimeDB identifies this connection, and a private access token which can be saved and later passed to [`withToken`](#method-withtoken) to authenticate the same user in future connections.

> This interface may change in an upcoming release as we rework SpacetimeDB's authentication model.

Reviewer: Will it? I was not aware of planned changes. But also I don't think we should say this, because this either saying it's not stable (even though we're claiming 1.0) or it's coming in 2.0 in which case we should just not say it.

Reviewer: Ah okay, that makes more sense.

#### Callback `onConnectError`

```typescript
DbConnectionBuilder.onConnectError(
    callback: (ctx: ErrorContext, error: Error) => void
): DbConnectionBuilder
```

Chain a call to `.onConnectError(callback)` to your builder to register a callback to run when your connection fails.

#### Callback `onDisconnect`

```typescript
DbConnectionBuilder.onDisconnect(
    callback: (ctx: ErrorContext, error: Error | null) => void
): DbConnectionBuilder

```

Chain a call to `.onDisconnect(callback)` to your builder to register a callback to run when your `DbConnection` disconnects from the remote module, either as a result of a call to [`disconnect`](#method-disconnect) or due to an error.

#### Method `withToken`

```typescript

> DbConnectionBuilder.withToken(token: string | null): DbConnectionBuilder

Reviewer: ```suggestion
DbConnectionBuilder.withToken(token: string | undefined): DbConnectionBuilder
```

JavaScript is a silly language. This is a meaningful distinction.

Author: I checked the SDK code when I wrote this, and was pretty sure it said `undefined`, not `null`. But now I'm looking again, and it seems to have been changed to (or always was) `token?: string`.

That said, based on [the TypeScript handbook](https://www.typescriptlang.org/docs/handbook/2/functions.html#optional-parameters), it looks like `token?: string` means `string | undefined`, *not* `string | null`.

Am I missing something here?

Reviewer: I also checked. `token?: string` does mean `string | undefined`. In your docs it's `string | null` which is not correct.

Author: Ah, I see. Will fix. I'm going to just write `token?: string`.

```

Chain a call to `.withToken(token)` to your builder to provide an OpenID Connect compliant JSON Web Token to authenticate with, or to explicitly select an anonymous connection. If this method is not called or `null` is passed, SpacetimeDB will generate a new `Identity` and sign a new private access token for the connection.

#### Method `build`

```typescript
DbConnectionBuilder.build(): DbConnection
```

After configuring the connection and registering callbacks, attempt to open the connection.

### Access tables and reducers

#### Field `db`

```rust
DbConnection.db: RemoteTables
```

The `db` field of the `DbConnection` provides access to the subscribed view of the remote database's tables. See [Access the client cache](#access-the-client-cache).

#### Field `reducers`

```rust
DbConnection.reducers: RemoteReducers
```

The `reducers` field of the `DbConnection` provides access to reducers exposed by the remote module. See [Observe and invoke reducers](#observe-and-invoke-reducers).

## Interface `DbContext`

```rust

> interface spacetimedb_sdk.DbContext<

Reviewer: ```suggestion
interface DbContext<
```

#### Field `db`

```typescript
DbConnection.db: DbView
```

The `db` field of a `DbContext` provides access to the subscribed view of the remote database's tables. See [Access the client cache](#access-the-client-cache).

#### Field `reducers`

```typescript
DbConnection.reducers: Reducers
```

The `reducers` field of a `DbContext` provides access to reducers exposed by the remote module. See [Observe and invoke reducers](#observe-and-invoke-reducers).

### Method `disconnect`

```typescript
DbContext.disconnect(): void
```

Gracefully close the `DbConnection`. Throws an error if the connection is already disconnected.

### Subscribe to queries

| Name                                                    | Description                                                 |
|---------------------------------------------------------|-------------------------------------------------------------|
| [`SubscriptionBuilder` type](#type-subscriptionbuilder) | Builder-pattern constructor to register subscribed queries. |
| [`SubscriptionHandle` type](#type-subscriptionhandle)   | Manage an active subscripion.                               |

#### Type `SubscriptionBuilder`

```typescript

> spacetimedb_sdk::SubscriptionBuilder

Reviewer: This type is actually in the module bindings, not the sdk IIRC.

```typescript
DbContext.subscriptionBuilder(): SubscriptionBuilder

```

Subscribe to queries by calling `ctx.subscription_builder()` and chaining configuration methods, then calling `.subscribe(queries)`.

##### Callback `onApplied`

```typescript
SubscriptionBuilder.onApplied(
    callback: (ctx: SubscriptionEventContext) => void
): SubscriptionBuilder
```

Register a callback to run when the subscription is applied and the matching rows are inserted into the client cache.

##### Callback `onError`

```typescript
SubscriptionBuilder.onError(
    callback: (ctx: ErrorContext, error: Error) => void
): SubscriptionBuilder

```

Register a callback to run if the subscription is rejected or unexpectedly terminated by the server. This is most frequently caused by passing an invalid query to [`subscribe`](#method-subscribe).

##### Method `subscribe`

```typescript
SubscriptionBuilder.subscribe(queries: string | string[]): SubscriptionHandle
```

Subscribe to a set of queries.

See [the SpacetimeDB SQL Reference](/docs/sql#subscriptions) for information on the queries SpacetimeDB supports as subscriptions.

##### Method `subscribeToAllTables`

```typescript

> SubscriptionBuilder.subscribe_to_all_tables(): void

Reviewer 2: ```suggestion
SubscriptionBuilder.subscribeToAllTables(): void
```

```

> Subscribe to all rows from all tables. This method is provided as a convenience for simple clients. The same connection must not mix `subscribeToAllTables` with [`subscribe` to specific queries](#method-subscribe). Doing so may cause errors or corrupt the client cache, leading to local queries returning incorrect results. The subscription initiated by `subscribeToAllTables` cannot be canceled after it is initiated.

Reviewer 2: ```suggestion
Subscribe to all rows from all public tables. This method is provided as a convenience for simple clients. The subscription initiated by `subscribeToAllTables ` cannot be canceled after it is initiated. You should [`subscribe` to specific queries](#method-subscribe) if you need fine-grained control over the lifecycle of your subscriptions.
```

Same change from the rust sdk docs.

#### Type `SubscriptionHandle`

```rust

> module_bindings::SubscriptionHandle

Reviewer: There is technically no such type in the TypeScript SDK. We just return: `SubscriptionHandleImpl<DBView, Reducers, SetReducerFlags>`

Reviewer: Not a huge deal though since it is that type in spirit.

Author: Does this mean that a user who wants to store a subscription handle in some sort of data structure has to either write out `SubscriptionHandleImpl<DBView, Reducers, SetReducerFlags>` or cast to `any`?

##### Method `isEnded`

```typescript
SubscriptionHandle.isEnded(): bool
```

Returns true if this subscription has been terminated due to an unsubscribe call or an error.

##### Method `isActive`

```typescript
SubscriptionHandle.isActive(): bool
```

Returns true if this subscription has been applied and has not yet been unsubscribed.

##### Method `unsubscribe`

```typescript
SubscriptionHandle.unsubscribe(): void
```

Terminate this subscription, causing matching rows to be removed from the client cache. Any rows removed from the client cache this way will have [`onDelete` callbacks](#callback-ondelete) run for them.

Unsubscribing is an asynchronous operation. Matching rows are not removed from the client cache immediately. Use [`unsubscribeThen`](#method-unsubscribethen) to run a callback once the unsubscribe operation is completed.

Throws an error if the subscription has already ended, either due to a previous call to `unsubscribe` or [`unsubscribeThen`](#method-unsubscribethen), or due to an error.

##### Method `unsubscribeThen`

```typescript

> SubscriptionHandle.unsubscribe_then(

Reviewer: ```suggestion
SubscriptionHandle.unsubscribeThen(
```

    on_end: (ctx: SubscriptionEventContext) => void
): void

```

Terminate this subscription, and run the `onEnd` callback when the subscription is ended and its matching rows are removed from the client cache. Any rows removed from the client cache this way will have [`onDelete` callbacks](#callback-ondelete) run for them.

> Returns an error if the subscription has already ended, either due to a previous call to [`unsubscribe`](#method-unsubscribe) or `unsubscribe_then`, or due to an error.

Reviewer: ```suggestion
Returns an error if the subscription has already ended, either due to a previous call to [`unsubscribe`](#method-unsubscribe) or `unsubscribeThen`, or due to an error.
```

): void;

TableHandle.removeOnDelete(
    callback: (ctx: EventContext, row: Row) => void
): void;
```

The `onDelete` callback runs whenever a previously-resident row is deleted from the client cache.

The `Row` type will be an autogenerated type which matches the row type defined by the module.

`removeOnDelete` may be used to un-register a previously-registered `onDelete` callback.

### Callback `onUpdate`

```typescript
TableHandle.onUpdate(
    callback: (ctx: EventContext, old: Row, new: Row) => void
): void;

TableHandle.removeOnUpdate(
    callback: (ctx: EventContext, old: Row, new: Row) => void
): void;
```

The `onUpdate` callback runs whenever an already-resident row in the client cache is updated, i.e. replaced with a new row that has the same primary key.

Only tables with a declared primary key expose `onUpdate` callbacks. Handles for tables without a declared primary key will not have `onUpdate` or `removeOnUpdate` methods.

The `Row` type will be an autogenerated type which matches the row type defined by the module.

`removeOnUpdate` may be used to un-register a previously-registered `onUpdate` callback.

### Unique constraint index access

> For each unique constraint on a table, its table handle has a field whose name is the unique column name. This field is a unique index handle. The unique index handle has a method `.find(desired_val: Col) -> Row | undefined`, where `Col` is the type of the column, and `Row` the type of rows. If a row with `desired_val` in the unique column is resident in the client cache, `.find` returns it.

Reviewer: ```suggestion
For each unique constraint on a table, its table handle has a field whose name is the unique column name. This field is a unique index handle. The unique index handle has a method `.find(desiredValue: Col) -> Row | undefined`, where `Col` is the type of the column, and `Row` the type of rows. If a row with `desiredValue` in the unique column is resident in the client cache, `.find` returns it.
```
